src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,235 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use Iterator;
+use IteratorAggregate;
+use MongoDB\BSON\Document;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONCodecLibrary;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_key_exists;
+use function assert;
+use function is_array;
+use function is_object;
+use function is_string;
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON document.
+ *
+ * The internal data will be cast to an object during BSON serialization to
+ * ensure that it becomes a BSON document.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<string, TValue>
+ * @template-implements IteratorAggregate<string, TValue>
+ */
+class LazyBSONDocument implements ArrayAccess, IteratorAggregate
+{
+    /** @var Document<TValue> */
+    private Document $bson;
+
+    /** @var array<string, TValue> */
+    private array $read = [];
+
+    /** @var array<string, bool> */
+    private array $exists = [];
+
+    /** @var array<string, TValue> */
+    private array $set = [];
+
+    /** @var array<string, true> */
+    private array $unset = [];
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy document.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /**
+     * Constructs a lazy BSON document.
+     *
+     * @param Document<TValue>|array<TValue>|object|null $input An input for a lazy object.
+     *        When given a BSON document, this is treated as input. For arrays
+     *        and objects this constructs a new BSON document using fromPHP.
+     */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = Document::fromPHP([]);
+        } elseif ($input instanceof Document) {
+            $this->bson = $input;
+        } elseif (is_array($input) || is_object($input)) {
+            $this->bson = Document::fromPHP([]);
+
+            foreach ($input as $key => $value) {
+                assert(is_string($key));
```

Comment:
You can specify an exception as 2nd argument.
```php
assert(is_string($key), InvalidArgumentException::...);
```

Reply:
In this case, the `assert` was there to silence Psalm. I worked around this by using `get_object_vars` for objects and specifying an `array<string, TValue>` type for arrays, which removes the need for asserting on the type.